src/stan/model/indexing/lvalue_varmat.hpp

```diff
@@ -0,0 +1,396 @@
+ *
+ * @tparam Vec `var_value` with inner Eigen type with either dynamic rows or
+ * columns, but not both.
+ * @tparam U Type of value (must be assignable to T).
+ * @param[in] x Vector variable to be assigned.
+ * @param[in] idxs index holding which cell to assign to.
+ * @param[in] y Value to assign.
+ * @param[in] name Name of variable (default "ANON").
+ * @param[in] depth Indexing depth (default 0).
+ * @throw std::out_of_range If the index is out of bounds.
+ */
+template <typename VarVec, typename U, require_var_vector_t<VarVec>* = nullptr,
+          require_var_t<U>* = nullptr,
+          require_floating_point_t<value_type_t<U>>* = nullptr>
+inline void assign(VarVec&& x,
+                   const cons_index_list<index_uni, nil_index_list>& idxs,
+                   const U& y, const char* name = "ANON", int depth = 0) {
+  stan::math::check_range("var_vector[uni] assign", name, x.size(),
+                          idxs.head_.n_);
+  const auto coeff_idx = idxs.head_.n_ - 1;
+  double prev_val = x.val().coeffRef(coeff_idx);
+  x.vi_->val_.coeffRef(coeff_idx) = y.val();
+  stan::math::reverse_pass_callback([x, y, coeff_idx, prev_val]() mutable {
+    x.vi_->val_.coeffRef(coeff_idx) = prev_val;
+    prev_val = x.adj().coeffRef(coeff_idx);
+    x.adj().coeffRef(coeff_idx) = 0.0;
+    y.adj() += prev_val;
+  });
+}
+
+/**
+ * Assign to a non-contiguous subset of elements in a vector.
+ *
+ * Types:  vector[multi] <- vector
+ *
+ * @tparam Vec1 `var_value` with inner Eigen type with either dynamic rows or
+ * columns, but not both.
+ * @tparam Vec2 `var_value` with inner Eigen type with either dynamic rows or
+ * columns, but not both.
+ * @param[in] x Vector to be assigned.
+ * @param[in] idxs Index holding an `std::vector` of cells to assign to.
+ * @param[in] y Value vector.
+ * @param[in] name Name of variable (default "ANON").
+ * @param[in] depth Indexing depth (default 0).
+ * @throw std::out_of_range If any of the indices are out of bounds.
+ * @throw std::invalid_argument If the value size isn't the same as
+ * the indexed size.
+ */
+template <typename Vec1, typename Vec2,
+          require_all_var_vector_t<Vec1, Vec2>* = nullptr>
+inline void assign(Vec1&& x,
+                   const cons_index_list<index_multi, nil_index_list>& idxs,
+                   const Vec2& y, const char* name = "ANON", int depth = 0) {
+  stan::math::check_size_match("vector[multi] assign", "left hand side",
+                               idxs.head_.ns_.size(), name, y.size());
+  const auto x_size = x.size();
+  const auto assign_size = idxs.head_.ns_.size();
+  arena_t<std::vector<int>> x_idx;
+  arena_t<std::vector<int>> y_idx;
+  // Need to remove duplicates for cases like {2, 3, 2, 2}
+  for (int i = assign_size - 1; i >= 0; --i) {
+    if (!internal::check_duplicate(x_idx, idxs.head_.ns_[i] - 1)) {
+      stan::math::check_range("vector[multi] assign", name, x_size,
+                              idxs.head_.ns_[i]);
+      y_idx.push_back(i);
```

Comment:
I'm not sure I'm following what you mean here, I just need `y_idx` for the duplicate index thing

Reply:
I mean you can completely remove the variable `y_idx` and replace all uses `y_idx[i]` with `i`.

Reply:
What about 

```stan
x[{1, 2, 2, 3}] = y
```

If we don't store the y index associated with the deduped x then we'll end up iterating over y[{4, 3, 2}] where we want to iterate over `y[{4, 3, 1}]`

Reply:
I see. How about my implementation:
```
template <typename Vec1, typename Vec2,
          require_all_var_vector_t<Vec1, Vec2>* = nullptr>
inline void assign(Vec1&& x,
                   const cons_index_list<index_multi, nil_index_list>& idxs,
                   const Vec2& y, const char* name = "ANON", int depth = 0) {
  stan::math::check_size_match("vector[multi] assign", "left hand side",
                               idxs.head_.ns_.size(), name, y.size());
  const auto x_size = x.size();
  const auto assign_size = idxs.head_.ns_.size();
  arena_t<std::vector<int>> x_idx(assign_size);
  arena_t<Eigen::Matrix<double, -1, 1>> prev_vals(assign_size);
  std::unordered_set<int> x_set();
  x_set.reserve(assign_size);
  for (int i = assign_size - 1; i >= 0; --i) {
    if (likely(x_set.insert(idxs.head_.ns_[i]).second)) {
      stan::math::check_range("vector[multi] assign", name, x_size,
                              idxs.head_.ns_[i]);
      x_idx[i] = idxs.head_.ns_[i] - 1;
      prev_vals.coeffRef(i) = x.vi_->val_.coeffRef(x_idx[i]);
      x.vi_->val_.coeffRef(x_idx[i]) = y.vi_->val_.coeff(i);
	}
	else{
	  x_idx[i] = -1;
	}
  }
  stan::math::reverse_pass_callback([x, y, x_idx, prev_vals]() mutable {
    for (Eigen::Index i = 0; i < x_idx.size(); ++i) {
	  if(likely(x_idx[i]! = -1)){
        x.vi_->val_.coeffRef(x_idx[i]) = prev_vals.coeffRef(i);
        prev_vals.coeffRef(i) = x.adj().coeffRef(x_idx[i]);
        x.adj().coeffRef(x_idx[i]) = 0.0;
        y.adj().coeffRef(i) += prev_vals.coeffRef(i);
	  }
    }
  });
}
```
This avoids `y_idx` and the use of `x_set` in reverse pass at expense of additional if in reverse pass. I believe this would be faster.

Reply:
Ah! I like that